### PR TITLE
C++: Remove TC from `Element.getRootCause`

### PR DESCRIPTION
**Pull request by MathiasVP**
Before:
```ql
Tuple counts for Element::Element::findRootCause#dispred#f0820431#ff#higher_order_body/1@2375156s after 28.6s:
  2507326   ~1%         {1} r1 = SCAN macroinvocations OUTPUT In.0
  79809321  ~508%       {1} r2 = JOIN r1 WITH Macro::MacroInvocation::getAGeneratedElement#dispred#f0820431#ff ON FIRST 1 OUTPUT Rhs.1
  424223876 ~21396%     {1} r3 = JOIN r2 WITH Macro::MacroInvocation::getAGeneratedElement#dispred#f0820431#ff_10#join_rhs ON FIRST 1 OUTPUT Rhs.1 'arg0'
                        return r3

Starting to evaluate predicate boundedFastTC(Macro::MacroAccess::getParentInvocation#dispred#f0820431#ff,Element::Element::findRootCause#dispred#f0820431#ff#higher_order_body)/2@b7e8f58u = HOP boundedFastTC(2263190,1949320)
 >>> Relation boundedFastTC(Macro::MacroAccess::getParentInvocation#dispred#f0820431#ff,Element::Element::findRootCause#dispred#f0820431#ff#higher_order_body): 10393246 rows using 68 MB

Tuple counts for Element::Element::findRootCause#dispred#f0820431#ff#antijoin_rhs/3@836dc149 after 2m18s:
  2507326   ~0%       {2} r1 = SCAN macroinvocations OUTPUT In.0 'arg0', In.1 'arg1'
  79809321  ~0%       {3} r2 = JOIN r1 WITH Macro::MacroInvocation::getAGeneratedElement#dispred#f0820431#ff ON FIRST 1 OUTPUT Rhs.1 'arg2', Lhs.0 'arg0', Lhs.1 'arg1'
  531344575 ~0%       {4} r3 = JOIN r2 WITH Macro::MacroInvocation::getAGeneratedElement#dispred#f0820431#ff_10#join_rhs ON FIRST 1 OUTPUT Rhs.1, Lhs.1 'arg0', Lhs.2 'arg1', Lhs.0 'arg2'
  225692847 ~236%     {3} r4 = JOIN r3 WITH boundedFastTC(Macro::MacroAccess::getParentInvocation#dispred#f0820431#ff,Element::Element::findRootCause#dispred#f0820431#ff#higher_order_body) ON FIRST 2 OUTPUT Lhs.1 'arg0', Lhs.2 'arg1', Lhs.3 'arg2'
                      return r4

Tuple counts for Element::Element::findRootCause#dispred#f0820431#ff/2@521a82id after 6.7s:
  79809321 ~0%     {3} r1 = JOIN macroinvocations WITH Macro::MacroInvocation::getAGeneratedElement#dispred#f0820431#ff ON FIRST 1 OUTPUT Lhs.0, Lhs.1 'result', Rhs.1 'this'
  13098880 ~0%     {3} r2 = r1 AND NOT Element::Element::findRootCause#dispred#f0820431#ff#antijoin_rhs(Lhs.0, Lhs.1 'result', Lhs.2 'this')
  13098880 ~0%     {2} r3 = SCAN r2 OUTPUT In.2 'this', In.1 'result'
                    return r3
```
After:
```ql
Tuple counts for Element::hasCloserMacroInvocation#496c7fc2#ff/2@f0d547uf after 11s:
  2315963  ~0%     {3} r1 = SCAN macroparent OUTPUT In.1 'mi', 1, In.0
  2263190  ~4%     {2} r2 = JOIN r1 WITH macroinvocations_03#join_rhs ON FIRST 2 OUTPUT Lhs.2, Lhs.0 'mi'
  66710441 ~0%     {2} r3 = JOIN r2 WITH Macro::MacroInvocation::getAGeneratedElement#dispred#f0820431#ff ON FIRST 1 OUTPUT Rhs.1 'elem', Lhs.1 'mi'
                    return r3
Registering Element::hasCloserMacroInvocation#496c7fc2#ff/2@f0d547uf +  with content 9742c74q3f6n0eq8m0rl6opo3f4
>>> Created relation Element::hasCloserMacroInvocation#496c7fc2#ff/2@f0d547uf with 66710441 rows.
Starting to evaluate predicate Element::Element::findRootCause#dispred#f0820431#ff/2@ea6d23et
Tuple counts for Element::Element::findRootCause#dispred#f0820431#ff/2@ea6d23et after 8.6s:
  79809321 ~0%     {3} r1 = JOIN macroinvocations WITH Macro::MacroInvocation::getAGeneratedElement#dispred#f0820431#ff ON FIRST 1 OUTPUT Lhs.0, Lhs.1 'result', Rhs.1 'this'
  13098880 ~0%     {3} r2 = r1 AND NOT Element::hasCloserMacroInvocation#496c7fc2#ff(Lhs.2 'this', Lhs.0)
  13098880 ~0%     {2} r3 = SCAN r2 OUTPUT In.2 'this', In.1 'result'
                    return r3
```

### CODE REVIEW

**Review by geoffw0 — APPROVED** [review]
LGTM.  Still not cheap, but many times better.

**Comment by MathiasVP** [comment]
> LGTM. Still not cheap, but many times better.

Indeed. Anything that joins with `getAGeneratedElement` is bound to be very large, unfortunately.

File: cpp/ql/lib/semmle/code/cpp/Element.qll [hunk 1]
```diff
@@ -236,6 +233,14 @@ class Element extends ElementBase {
   }
 }
 
+pragma[noinline]
+private predicate hasCloserMacroInvocation(Element elem, MacroInvocation mi) {
+  exists(MacroInvocation closer |
+    elem = closer.getAGeneratedElement() and
+    mi = closer.getParentInvocation()
```
[the remark this answers is not in the document]
Comment (MathiasVP) [reply cpp/ql/lib/semmle/code/cpp/Element.qll:240]: Thanks for highlighting this, yes. I ran that query on an additional ~120 of our favorite projects and also came back with no differences.

**Comment by MathiasVP** [comment]
It's difficult to say whether this actually fixes the OOM since we're currently OOM'ing on the `main` branch and so we never actually run this PR on the problematic project. But this looks like an improvement locally so I'm just gonna go ahead and merge this.

**Pull request merged**